File: test_chapters.js
// Script para probar la extracción de capítulos en la consola del navegador
// Abre https://libribar.com/manga/la-leyenda-de-la-estrella-general/ y pega este script en la consola

console.log("=== INICIANDO TEST DE EXTRACCIÓN DE CAPÍTULOS ===");

// 1. Buscar todos los elementos li.wp-manga-chapter
const chapterLis = document.querySelectorAll("li.wp-manga-chapter");
console.log(`✓ Elementos <li class="wp-manga-chapter"> encontrados: ${chapterLis.length}`);

if (chapterLis.length === 0) {
    console.error("❌ NO SE ENCONTRARON ELEMENTOS <li> CON LA CLASE 'wp-manga-chapter'");
    console.log("Clases disponibles en los <li>:");
    document.querySelectorAll("li").forEach((li, i) => {
        if (i < 5) console.log(`  - ${li.className}`);
    });
} else {
    console.log("\n--- ANALIZANDO PRIMER CAPÍTULO ---");
    const firstLi = chapterLis[0];

    // Mostrar HTML del primer elemento
    console.log("HTML del primer <li>:");
    console.log(firstLi.outerHTML.substring(0, 500));

    // 2. Intentar buscar con el selector: span.chapter-link-container a
    console.log("\n--- PROBANDO SELECTOR: 'span.chapter-link-container a' ---");
    const linkWithContainer = firstLi.querySelector("span.chapter-link-container a");

    if (linkWithContainer) {
        console.log("✓ Enlace encontrado con 'span.chapter-link-container a'");
        console.log(`  - href: ${linkWithContainer.getAttribute('href')}`);
        console.log(`  - texto: "${linkWithContainer.textContent.trim()}"`);
        console.log(`  - clase: ${linkWithContainer.className}`);
    } else {
        console.error("❌ NO se encontró enlace con 'span.chapter-link-container a'");
    }

    // 3. Probar fallback: buscar cualquier <a>
    console.log("\n--- PROBANDO FALLBACK: cualquier 'a' ---");
    const anyLink = firstLi.querySelector("a");

    if (anyLink) {
        console.log("✓ Enlace encontrado con selector genérico 'a'");
        console.log(`  - href: ${anyLink.getAttribute('href')}`);
        console.log(`  - texto: "${anyLink.textContent.trim()}"`);
        console.log(`  - clase: ${anyLink.className}`);
    } else {
        console.error("❌ NO se encontró ningún enlace <a>");
    }

    // 4. Buscar todos los <a> para ver si hay múltiples
    console.log("\n--- TODOS LOS ENLACES EN EL PRIMER <li> ---");
    const allLinks = firstLi.querySelectorAll("a");
    console.log(`Total de enlaces <a>: ${allLinks.length}`);
    allLinks.forEach((link, index) => {
        console.log(`  Enlace ${index + 1}:`);
        console.log(`    - href: ${link.getAttribute('href')}`);
        console.log(`    - texto: "${link.textContent.trim()}"`);
        console.log(`    - clase: ${link.className}`);
        console.log(`    - parent: ${link.parentElement.className}`);
    });

    // 5. Extraer TODOS los capítulos
    console.log("\n=== EXTRAYENDO TODOS LOS CAPÍTULOS ===");
    let successCount = 0;
    let failCount = 0;
    const chapters = [];

    chapterLis.forEach((li, index) => {
        const link = li.querySelector("span.chapter-link-container a") || li.querySelector("a");

        if (link && link.getAttribute('href')) {
            const chapterData = {
                index: index,
                name: link.textContent.trim(),
                url: link.getAttribute('href'),
                fullUrl: link.href
            };
            chapters.push(chapterData);
            successCount++;

            // Mostrar solo los primeros 5 para no saturar la consola
            if (index < 5) {
                console.log(`✓ Capítulo ${index + 1}: ${chapterData.name} -> ${chapterData.url}`);
            }
        } else {
            failCount++;
            if (index < 5) {
                console.error(`❌ Capítulo ${index + 1}: NO SE ENCONTRÓ ENLACE`);
            }
        }
    });

    console.log(`\n--- RESUMEN ---`);
    console.log(`✓ Capítulos extraídos exitosamente: ${successCount}`);
    console.log(`❌ Capítulos que fallaron: ${failCount}`);
    console.log(`📊 Total: ${chapterLis.length}`);

    // 6. Verificar si hay fechas
    console.log("\n--- PROBANDO EXTRACCIÓN DE FECHAS ---");
    const firstChapterLi = chapterLis[0];
    const dateWithTitle = firstChapterLi.querySelector("span.chapter-release-date a[title]");
    const dateWithI = firstChapterLi.querySelector("span.chapter-release-date i");
    const dateSpan = firstChapterLi.querySelector("span.chapter-release-date");

    console.log("Fecha con a[title]:", dateWithTitle ? dateWithTitle.getAttribute('title') : "NO ENCONTRADO");
    console.log("Fecha con <i>:", dateWithI ? dateWithI.textContent : "NO ENCONTRADO");
    console.log("Fecha en span:", dateSpan ? dateSpan.textContent.trim() : "NO ENCONTRADO");

    // 7. Guardar en variable global para inspección
    window.extractedChapters = chapters;
    console.log("\n✓ Capítulos guardados en: window.extractedChapters");
    console.log("  Puedes inspeccionarlos escribiendo: extractedChapters");
}

console.log("\n=== TEST COMPLETADO ===");

File: src/es/barmanga/src/eu/kanade/tachiyomi/extension/es/barmanga/TestChapters.kt
package eu.kanade.tachiyomi.extension.es.barmanga

import org.jsoup.Jsoup

/**
 * Test simple para verificar la extracción de capítulos
 *
 * Para ejecutar:
 * 1. Abre este archivo en Android Studio
 * 2. Click derecho en main() -> Run 'MainKt'
 *
 * O desde terminal:
 * gradlew.bat :src:es:barmanga:run (si está configurado)
 */

fun main() {
    println("=== TEST DE EXTRACCIÓN DE CAPÍTULOS ===\n")

    val mangaUrl = "https://libribar.com/manga/la-leyenda-de-la-estrella-general/"

    println("Conectando a: $mangaUrl")

    try {
        // Hacer la petición HTTP
        val document = Jsoup.connect(mangaUrl)
            .userAgent("Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36")
            .header("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,*/*;q=0.8")
            .header("Accept-Language", "es-ES,es;q=0.9,en;q=0.8")
            .get()

        println("✓ Página cargada exitosamente\n")

        // 1. Buscar los elementos <li class="wp-manga-chapter">
        val chapterListSelector = "li.wp-manga-chapter"
        val chapterElements = document.select(chapterListSelector)

        println("--- ANÁLISIS INICIAL ---")
        println("Selector usado: '$chapterListSelector'")
        println("Elementos <li> encontrados: ${chapterElements.size}\n")

        if (chapterElements.isEmpty()) {
            println("❌ NO SE ENCONTRARON CAPÍTULOS")
            println("\nBuscando alternativas...")

            val allLis = document.select("li")
            println("Total de <li> en la página: ${allLis.size}")

            if (allLis.isNotEmpty()) {
                println("\nPrimeras 5 clases de <li>:")
                allLis.take(5).forEach { li ->
                    println("  - ${li.className()}")
                }
            }

            return
        }

        // 2. Analizar el primer capítulo en detalle
        println("--- ANÁLISIS DEL PRIMER CAPÍTULO ---")
        val firstChapter = chapterElements.first()!!
        println("HTML del primer <li>:")
        println(firstChapter.html().take(500))
        println()

        // 3. Probar el selector específico (NUEVO: data-href en span)
        println("--- PROBANDO SELECTOR: 'span.chapter-link-inner[data-href]' ---")
        val linkWithDataHref = firstChapter.selectFirst("span.chapter-link-inner[data-href]")

        if (linkWithDataHref != null) {
            println("✓ Enlace encontrado con data-href")
            println("  - data-href: ${linkWithDataHref.attr("data-href")}")
            println("  - texto: '${linkWithDataHref.text().trim()}'")
            println("  - clase: ${linkWithDataHref.className()}\n")
        } else {
            println("❌ NO se encontró con 'span.chapter-link-inner[data-href]'\n")

            // Probar fallback antiguo (por si acaso)
            println("--- PROBANDO FALLBACK: buscar <a> tradicionales ---")
            val fallbackLinks = firstChapter.select("a")
            println("Total de enlaces <a>: ${fallbackLinks.size}")

            if (fallbackLinks.isNotEmpty()) {
                fallbackLinks.forEachIndexed { index, link ->
                    println("  Enlace ${index + 1}:")
                    println("    - href: ${link.attr("href")}")
                    println("    - texto: '${link.text().trim()}'")
                    println("    - clase: ${link.className()}")
                }
            }
            println()
        }

        // 4. Extraer TODOS los capítulos
        println("--- EXTRAYENDO TODOS LOS CAPÍTULOS ---")

        var successCount = 0
        var failCount = 0
        val chapters = mutableListOf<ChapterData>()

        chapterElements.forEachIndexed { index, element ->
            // NUEVO: Buscar span con data-href
            val linkElement = element.selectFirst("span.chapter-link-inner[data-href]")
            val link = linkElement?.attr("data-href")
            val title = linkElement?.text()?.trim().orEmpty()

            if (!link.isNullOrEmpty()) {
                val chapterData = ChapterData(
                    index = index,
                    name = title.ifEmpty { "Capítulo ${index + 1}" },
                    url = link,
                    absUrl = link, // ya es absoluto
                )
                chapters.add(chapterData)
                successCount++

                // Mostrar solo los primeros 5
                if (index < 5) {
                    println("✓ Capítulo ${index + 1}: ${chapterData.name} -> ${chapterData.url}")
                }
            } else {
                failCount++
                if (index < 5) {
                    println("❌ Capítulo ${index + 1}: NO se encontró enlace válido")
                }
            }
        }

        println("\n--- RESUMEN ---")
        println("✓ Capítulos extraídos exitosamente: $successCount")
        println("❌ Capítulos que fallaron: $failCount")
        println("📊 Total: ${chapterElements.size}\n")

        // 5. Probar extracción de fechas del primer capítulo
        println("--- EXTRACCIÓN DE FECHAS ---")
        val dateWithTitle = firstChapter.selectFirst("span.chapter-release-date a[title]")
        val dateWithI = firstChapter.selectFirst("span.chapter-release-date i")
        val dateSpan = firstChapter.selectFirst("span.chapter-release-date")

        println("Fecha con a[title]: ${dateWithTitle?.attr("title") ?: "NO ENCONTRADO"}")
        println("Fecha con <i>: ${dateWithI?.text() ?: "NO ENCONTRADO"}")
        println("Fecha en span: ${dateSpan?.text()?.trim() ?: "NO ENCONTRADO"}\n")

        // 6. Verificar si hay contenedor AJAX (para debug)
        println("--- VERIFICANDO CONTENEDOR AJAX ---")
        val ajaxContainer = document.select("div[id^=manga-chapters-holder]")
        println("Contenedor 'div[id^=manga-chapters-holder]' encontrado: ${ajaxContainer.size}")
        if (ajaxContainer.isNotEmpty()) {
            ajaxContainer.forEach { div ->
                println("  - id: ${div.id()}")
                println("  - data-id: ${div.attr("data-id")}")
            }
        }
        println()

        // 7. Mostrar estadísticas finales
        println("=== RESULTADO FINAL ===")
        if (successCount == chapterElements.size) {
            println("✅ ÉXITO TOTAL: Todos los capítulos extraídos correctamente")
        } else if (successCount > 0) {
            println("⚠️ PARCIAL: $successCount/${chapterElements.size} capítulos extraídos")
        } else {
            println("❌ FALLO: No se pudo extraer ningún capítulo")
        }
        println()

        // Mostrar los últimos 3 capítulos para verificar
        if (chapters.isNotEmpty()) {
            println("Últimos 3 capítulos:")
            chapters.takeLast(3).forEach { chapter ->
                println("  - ${chapter.name} -> ${chapter.url}")
            }
        }
    } catch (e: Exception) {
        println("❌ ERROR: ${e.message}")
        e.printStackTrace()
    }
}

data class ChapterData(
    val index: Int,
    val name: String,
    val url: String,
    val absUrl: String,
)
